integration-tests/downscope/downscope_test.go

```diff
@@ -0,0 +1,150 @@
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if testing.Short() {
+		// This line runs all of our individual tests
+		os.Exit(m.Run())
+	}
+	ctx := context.Background()
+	credentialFileName := os.Getenv(envServiceAccountFile)
+	credentialsFileData, err := os.ReadFile(credentialFileName)
+	if err != nil {
+		log.Fatalf("failed to open credentials file: %v", err)
+	}
+	rootCredential, err = google.CredentialsFromJSON(ctx, credentialsFileData, rootTokenScope)
+	if err != nil {
+		log.Fatalf("failed to construct root credential: %v", err)
+	}
+
+	// This line runs all of our individual tests
+	os.Exit(m.Run())
+
+}
+
+// downscopeTest holds the parameters necessary for running a test of the token downscoping capabilities implemented in `oauth2/google/downscope`
+type downscopeTest struct {
+	name                 string
+	availableResource    string
+	availablePermissions []string
+	conditions           []downscope.AvailabilityCondition
+	objectName           string
+	rootSource           oauth2.TokenSource
+	expectError          bool
+}
+
+func TestDownscopedToken(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	var downscopeTests = []downscopeTest{
+		{
+			name:                 "successfulDownscopedRead",
+			availableResource:    "//storage.googleapis.com/projects/_/buckets/" + bucket1,
+			availablePermissions: []string{"inRole:roles/storage.objectViewer"},
+			conditions: []downscope.AvailabilityCondition{
+				{
+					Expression: "resource.name.startsWith('projects/_/buckets/" + bucket1 + "/objects/" + object1 + "')",
+				},
+			},
+			rootSource:  rootCredential.TokenSource,
+			objectName:  object1,
+			expectError: false,
+		},
+		{
+			name:                 "readOWithoutPermission",
+			availableResource:    "//storage.googleapis.com/projects/_/buckets/" + bucket1,
+			availablePermissions: []string{"inRole:roles/storage.objectViewer"},
+			conditions: []downscope.AvailabilityCondition{
+				{
+					Expression: "resource.name.startsWith('projects/_/buckets/" + bucket1 + "/objects/" + object1 + "')",
+				},
+			},
+			rootSource:  rootCredential.TokenSource,
+			objectName:  object2,
+			expectError: true,
+		},
+	}
+
+	for _, tt := range downscopeTests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := helper(tt)
+			// If a test isn't supposed to fail, it shouldn't fail.
+			if !tt.expectError && err != nil {
+				t.Errorf("test case %v should have succeeded, but instead returned %v", tt.name, err)
+			} else if tt.expectError && err == nil { // If a test is supposed to fail, it should return a non-nil error.
+				t.Errorf(" test case %v should have returned an error, but instaed returned nil", tt.name)
+			}
+		})
+	}
+}
+
+// I'm not sure what I should name this according to convention.
+func helper(tt downscopeTest) error {
+	ctx := context.Background()
+
+	// Initializes an accessBoundary
+	AccessBoundaryRules := make([]downscope.AccessBoundaryRule, 0)
```

Comment:
nit: slices don't need to be initialized. `var AccessBoundaryRules []downscope.AccessBoundaryRule`